src/client/sandbox/cookie/server-sync.js

```diff
@@ -0,0 +1,122 @@
+import nativeMethods from '../../sandbox/native-methods';
+import XhrSandbox from '../../sandbox/xhr';
+import settings from '../../settings';
+import { stringify as stringifyJSON } from '../../json';
+import { isIE9 } from '../../utils/browser';
+
+const SWITCH_BACK_TO_ASYNC_XHR_DELAY = 2000;
+
+export default class ServerSync {
+    constructor () {
+        this.useAsyncXhr = true;
+        this.queue       = [];
+        this.sendedQueue = null;
+        this.activeReq   = null;
+
+        // NOTE: When unloading, we should switch to synchronous XHR to be sure that we won't lose any cookies.
+        nativeMethods.windowAddEventListener.call(window, 'beforeunload', () => this._beforeUnloadHandler(), true);
+    }
+
+    static _createXMLHttpRequest (isAsync) {
+        var xhr = XhrSandbox.createNativeXHR();
+
+        xhr.open('POST', settings.get().cookieSyncUrl, isAsync);
+        xhr.setRequestHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
+
+        return xhr;
+    }
+
+    static _isNoContentStatus (request) {
+        try {
+            // NOTE: XMLHTTPRequest implementation in MSXML HTTP does not handle HTTP responses
+            // with status code 204 (No Content) properly; the `status' property has the value 1223.
+            return request.status === (isIE9 ? 1223 : 204);
+        }
+        catch (e) {
+            return false;
+        }
+    }
+
+    _beforeUnloadHandler () {
+        if (this.useAsyncXhr) {
+            this.useAsyncXhr = false;
+
+            if (this.activeReq)
+                this.activeReq.abort();
+        }
+
+        // NOTE: If the unloading was canceled, switch back to asynchronous XHR.
+        nativeMethods.setTimeout.call(window, () => {
+            this.useAsyncXhr = true;
+        }, SWITCH_BACK_TO_ASYNC_XHR_DELAY);
+    }
+
+    _loadHandler (request) {
```

Comment:
Lets' rename to `onRequestLoad` and `onRequestError` to differ with page handler `_beforeUnloadHandler`